### PR TITLE
[Coverage][MCDC] Adapt mcdc to llvm 19

### PR DESCRIPTION
**Pull request by ZhuUx**
Related issue: #126672 

Also finish task 4 at #124144 

[llvm #82448](https://github.com/llvm/llvm-project/pull/82448) has introduced some break changes into mcdc, causing incompatibility between llvm 18 and 19. This draft adapts to that change and gives up supporting for llvm-18.

### CODE REVIEW

**Comment by ZhuUx** [comment]
It is still unfinished. I would add some options next days and try to warn about mcdc limits with lints as this [comment](https://github.com/rust-lang/rust/pull/126677#issuecomment-2178370336) suggest.

**Comment by ZhuUx** [comment]
I think we need to discuss more about mcdc options. Thus now I remove the commit and just reset `max_conditions_in_decision` to `i16::MAX` as default of clang.

Nevertheless I would wait until rustc merges llvm 19 and see if there are any other changes need to make.

**Comment by ZhuUx** [comment]
Now this pr is independent from #126677 . We can focus on this only.
@compiler-errors @nikic 
It's prepared for reviewing.
@rustbot review

File: compiler/rustc_mir_build/src/build/coverageinfo/mcdc.rs [hunk 1]
```diff
@@ -150,44 +155,64 @@ impl MCDCState {
         decision_ctx.decision_stack.push_back(lhs);
     }
 
-    fn take_condition(
+    fn try_finish_decision(
         &mut self,
+        span: Span,
         true_marker: BlockMarkerId,
         false_marker: BlockMarkerId,
-    ) -> (Option<ConditionInfo>, Option<MCDCDecisionSpan>) {
```
Comment (ZhuUx) [inline compiler/rustc_mir_build/src/build/coverageinfo/mcdc.rs:157]: This change looks a bit awkward. It's just a temporary expedient and will be removed in refactored mcdc implementation for pattern matching.

File: compiler/rustc_middle/src/mir/coverage.rs [hunk 2]
```diff
@@ -311,10 +297,9 @@ pub struct MCDCBranchSpan {
     pub span: Span,
     /// If `None`, this actually represents a normal branch span inserted for
     /// code that was too complex for MC/DC.
```
Comment (Zalathar) [inline compiler/rustc_middle/src/mir/coverage.rs:299]: This comment seems inaccurate now; should it be removed?

Comment (ZhuUx) [reply compiler/rustc_middle/src/mir/coverage.rs]: Sure it is removed at [compare](https://github.com/rust-lang/rust/compare/541e9a280ad3c73ebdd5230d3ec1f1af5a83d8b6..473131a57bc6cffd88272a22b384a9f08b46ab16)

**Comment by Zalathar** [comment]
Looks reasonable to me.

I haven't reviewed the MC/DC-specific parts in detail, but if the tests still pass then I assume things are no worse than they were before.

**Comment by Zalathar** [comment]
r? Zalathar

**Comment by Lambdaris** [comment]
:confounded: Nearly caused an accident, thankfully github is robust enough to restore.

I have not found a method to generate i16::MAX conditions without proc macros, so just check if it allows decisions containing more than 6 conditions. (Manually code so many conditions looks like an attack)

File: tests/coverage/mcdc/condition-limit.rs [hunk 3]
```diff
@@ -1,33 +1,18 @@
...
-    let [a, b, c, d, e, f] = <[bool; 6]>::default();
-    if a && b && c && d && e && f {
-        core::hint::black_box("hello");
-    }
-}
-
-fn bad() {
-    // With 7 conditions, fall back to branch instrumentation only.
-    let [a, b, c, d, e, f, g] = <[bool; 7]>::default();
+fn accpet_7_conditions(bool_arr: [bool; 7]) {
```
Comment (Zalathar) [inline tests/coverage/mcdc/condition-limit.rs:7]: Typo: `accpet` → `accept`

Comment (ZhuUx) [reply tests/coverage/mcdc/condition-limit.rs]: Thanks, fixed it

**Comment by Zalathar** [comment]
This needs another rebase because #131325 replaced the `bcb_counter` method with `term_for_bcb`.

Also please squish the “fixup” changes into the other patches as appropriate.

@rustbot author

**Pull request merged**